### PR TITLE
AP-4575: Add copy case page flows

### PR DESCRIPTION
**Pull request by jsugarman**
## What
Add copy case feature

[Link to story](https://dsdmoj.atlassian.net/browse/AP-4575)

The ability to copy case details from one application to another
is considered a prerequisite for the linking cases feature.

This branch adds the pages flows but does not do a full clone/copy of the necessary case details - see [AP-4577](https://dsdmoj.atlassian.net/browse/AP-4577) for that. Neither does it add the CYA page changes - see [AP-4601](https://dsdmoj.atlassian.net/browse/AP-4601) for that

## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

[AP-4577]: https://dsdmoj.atlassian.net/browse/AP-4577?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AP-4601]: https://dsdmoj.atlassian.net/browse/AP-4601?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: spec/requests/providers/copy_case_invitations_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      it "sets the application as draft" do
+        expect { patch_request }.to change { legal_aid_application.reload.draft? }.from(false).to(true)
+      end
+
+      it "does not record the answer" do
+        expect { patch_request }.not_to change { legal_aid_application.reload.copy_case }.from(nil)
+      end
+    end
+
+    # OPTIMISE: defined in a fair few places now, share?!
```
Comment (colinbruce) [inline spec/requests/providers/copy_case_invitations_controller_spec.rb:121]: Did you not do this already? Off the back of this?

File: spec/support/shared_context/session_double_shared_context.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    allow(session_double).to receive(:fetch) do |key|
+      session_hash.fetch(key)
+    end
+
+    allow(session_double).to receive(:key?) do |key|
+      session_hash.key?(key)
+    end
+
+    # rubocop:disable RSpec/AnyInstance
```
Comment (colinbruce) [inline spec/support/shared_context/session_double_shared_context.rb:34]: <squints>

**Review by colinbruce — COMMENTED** [review]
Looks good, a couple of observations from a linked-cases non-expert!

File: spec/requests/providers/copy_case_invitations_controller_spec.rb (hunk 1 above)
Comment (jsugarman) [reply spec/requests/providers/copy_case_invitations_controller_spec.rb:121]: yeah, need to rebase

File: spec/support/shared_context/session_double_shared_context.rb (hunk 2 above)
Comment (jsugarman) [reply spec/support/shared_context/session_double_shared_context.rb]: removed as session helper not needed anymore. although this was the best method I found for mocking session setting in request specs

**Comment by jsugarman** [comment]
> An observation from outside the team... Could the copy_case controllers, views, etc be namespaced into a folder to prevent the provider namespace bloating? e.g. `app/controllers/providers/copy_case/searches`, etc ? It helped with Partner work that we could find a section easier as it was in it's own namespaced folder :)

This is done now. see [commit](https://github.com/ministryofjustice/laa-apply-for-legal-aid/pull/5932/commits/eb296203dea73e702e49e7fe747a2b974f09e1a4)

**Review by kmahern — APPROVED** [review]
lgtm 👍

**Review by agoldstone93 — COMMENTED** [review]
Just one comment

File: app/controllers/providers/copy_case/searches_controller.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      def update
+        @form = ::CopyCase::SearchForm.new(form_params)
+
+        render :show unless save_continue_or_draft(@form)
+      end
+
+    private
+
+      def save_continue_or_draft(form, **)
+        draft_selected? ? form.save_as_draft : form.save!
```
[the remark this answers is not in the document]
Comment (jsugarman) [reply app/controllers/providers/copy_case/searches_controller.rb]: good spot. must have left it in or amend along the way to be exactly as in superclass.

[the remark this answers is not in the document]
Comment (jsugarman) [reply app/controllers/providers/copy_case/searches_controller.rb]: removed now

**Pull request merged**